src/aspire/utils/bot_align.py

```diff
@@ -0,0 +1,181 @@
+    :param tau: Regularization parameter for surrogate problems. Numeric, defaults 1e-3.
+    :param man_max_iter: Stopping criterion for surrogate problems--maximum iterations. Interger, defaults 500.
+    :param man_min_grad: Stopping criterion for surrogate problems--minimum gradient norm. Numeric, defaults 0.1.
+    :param man_min_sz: Stopping criterion for surrogate problems--minimum step size. Numeric, defaults 0.1.
+    :param verbosity: Whether to print surrogate problem optimization details. Boolean, defaults 0.
+    :param dtype: Numeric dtype to perform computations with.
+        Default `None` infers dtype from `vol_ref`.
+    :return: two rotation matrices R_init (without refinement) and R_est (with refinement).
+        If refine = False, then R_est is set equal to R_init.
+    """
+    # Infer dtype
+    dtype = np.dtype(dtype or vol_ref.dtype)
+
+    # Convert volume data dtype if needed.
+    vol_ref = vol_ref.astype(dtype, copy=False)
+    vol_given = vol_given.astype(dtype, copy=False)
+
+    # Store parameters specific to each loss_type.
+    LOSS_TYPES = {
+        "wemd": dict(lengthscale=0.75, wavelet="sym3", level=6),
+        "eu": dict(lengthscale=1),
+    }
+    loss_type = loss_type.lower()
+    if loss_type not in LOSS_TYPES.keys():
+        raise ValueError(f"BOTalign `loss_type` must be one of {LOSS_TYPES.keys()}")
+
+    # Lookup our params
+    loss_params = LOSS_TYPES[loss_type]
+
+    # Downsample Volumes
+    vol_ref_ds = vol_ref.downsample(downsampling_level)
+    vol_given_ds = vol_given.downsample(downsampling_level)
+
+    if loss_type == "wemd":
+        embed_0 = wemd_embed(
+            vol_ref_ds.asnumpy()[0], loss_params["wavelet"], loss_params["level"]
+        )
+
+    def loss(R):
+        """
+        Alignment loss function.
+        """
+        v = vol_given_ds.rotate(Rotation(R)).asnumpy()[0]
+        if loss_type == "eu":
+            return norm(v - vol_ref_ds)
+        elif loss_type == "wemd":
+            embed_v = wemd_embed(v, loss_params["wavelet"], loss_params["level"])
+            return norm(embed_v - embed_0, ord=1)
+
+    def cf(x1, x2):
+        """
+        The squared exponential covariance function over SO(3).
+        """
+        d = norm(x1 - x2) / loss_params["lengthscale"]
+        return np.exp(-(d**2) / 2, dtype=dtype)
+
+    def cf_grad(x1, x2):
+        """
+        Gradient of the squared exponential covariance function.
+        """
+        return cf(x1, x2) * (x2 - x1) / loss_params["lengthscale"] ** 2
+
+    R = np.zeros((max_iters, 3, 3), dtype=dtype)
+    R[0] = np.eye(3, dtype=dtype)
+    manifold = pymanopt.manifolds.SpecialOrthogonalGroup(3)
+
+    C = np.zeros((max_iters, max_iters), dtype=dtype)
+    C[0, 0] = cf(R[0], R[0])
+
+    y = np.zeros(max_iters, dtype=dtype)
+    y[0] = loss(R[0])
+
+    for t in range(1, max_iters):
+        q = np.linalg.solve(C[:t, :t] + tau * np.eye(t, dtype=dtype), y[:t])
```

Comment:
Brief explanation of what this does and why?

Reply:
@RuiyiYang , can you comment on this? Thanks

Reply:
Added comment relating to paper.